dask/array/slicing.py

```diff
@@ -697,6 +693,72 @@ def replace_ellipsis(n, index):
             index[loc + 1:])
 
 
+def normalize_slice(idx, dim):
+    """ Normalize slices to canonical form
+
+    Parameters
+    ----------
+    idx: slice or other index
+    dim: dimension length
+
+    Examples
+    --------
+    >>> normalize_slice(slice(0, 10, 1), 10)
+    slice(None, None, None)
+    """
+
+    if isinstance(idx, slice):
+        start, stop, step = idx.start, idx.stop, idx.step
+        if start is not None:
+            while start < 0:
+                start += dim
+        if stop is not None:
+            while stop < 0:
+                stop += dim
```

Comment:
I don't think this is correct. This seems to imply that given dim 10, stop -11 should be equal to 9, which isn't true. I think you want to just add `dim` once, and zero if still negative.

```python
In [7]: import numpy as np

In [8]: x = np.arange(10)

In [9]: x[:-1]
Out[9]: array([0, 1, 2, 3, 4, 5, 6, 7, 8])

In [10]: x[:-11]
Out[10]: array([], dtype=int64)

In [11]: x[:9]
Out[11]: array([0, 1, 2, 3, 4, 5, 6, 7, 8])
```

Reply:
Thanks.  I've stolen your code here and added the appropriate tests